### PR TITLE
Suggesting changes to CodingStyle.md

### PR DESCRIPTION
**Pull request by cpsw**
I would like to suggest to include some additional information on Code Style. This is a new direction towards avoiding certain patterns of code that might increase the changes of adding runtime errors, in my opinion. At the same time, I would invite others to share other patterns like these ones with goal of creating a set of guidelines that we could share among open-source projects.

### CODE REVIEW

**Comment by mikeller** [comment]
Thanks for suggesting this, @cpsw. I fully agree with the first two statements.

About the third one I am not sure. My preferred rule for this is `use only the parentheses that are required`. The reason for this is that it eliminates judgement calls ('is the precedence here obvious or does it need parentheses'), and the flip-flop churn that is associated with judgement calls, people falling on both sides of the answer, and large development teams.
The alternative way to eliminate judgement calls, 'use parentheses for every operator', becomes unwieldy very quickly.
Thoughts?

**Comment by martinbudden** [comment]
I agree with the first two, but not the third. Non-required parenthesis make the code less readable and in fact are distracting. If you see a parenthesis then the thought "ah something unusual is going on here" occurs and you then have to spend cognitive energy to figure out that nothing unusual is actually happening.

However, I think that parenthesis should be used for bitwise `&` and `|` operators since the precedence of these is problematic - however the compiler warns for those cases.

**Comment by McGiverGim** [comment]
I prefer to have parenthesis always, to me it makes simpler to read. But I think I'm alone with this 🤗

**Comment by cpsw** [comment]
Thanks for the answers!

The reason for suggesting the third rule is that by adding parentheses to the && expression, you make it explicit that the second part of the expression is going to be evaluated first. In cases that we have expressions with side effects, we might have problems with that.. For experienced developers, it is not a big issue, but it is easy to go unnoticed and cause runtime errors..

@martinbudden, what do you mean by saying that bitwise & and | operators are problematic?

**Comment by martinbudden** [comment]
The bitwise operators have lower precedence than the comparison operators, so if you do:
```
if (a & 1 ==0) {
...
```
you will get a warning `warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]`

**Comment by martinbudden** [comment]
@cpsw wrt `&&` and `||` I don't think parenthesis are any more required than they are for `*` and `+`.

and I certainly would be against the following parentheses:
```
if ( var1 + (var2 * var3) ) {
	someMethod( var1 );
}
```

**Comment by cpsw** [comment]
Thanks, @martinbudden! That's make totally sense. 

And what about the place to put this information, rules number 1 and 2. I will remove rule number 3. Is this file (CodingStyle.md) the right place for that? Are you all okay to include this information in the project?

Please let me know. Thanks

**Comment by cpsw** [comment]
I also would like to bring more patterns like these ones for discussion in subsequent pull requests, if you are interested. Regards!

File: docs/development/CodingStyle.md [hunk 1]
```diff
@@ -257,3 +257,34 @@ It should be based on sound mathematical, physical or computer science principle
 This is important for test code too. Tests shall be based on such principles and real-world properties so they don't just test the current implementation as it happens to be.
 
 Guidelines not tramlines: guidelines are not totally rigid - they can be broken when there is good reason.
+
+#Avoiding Accidental Mistakes
+
+* Always use braces when nesting if/else statements to avoid dangling else (the else branch belongs to the innermost if statement).
+
+```cpp
+if ( var1 ) {
```
Comment (ledvinap) [inline docs/development/CodingStyle.md]: Only space between keyword and parenhesis:
```
if (var1) {

File: docs/development/CodingStyle.md [hunk 2]
```diff
@@ -257,3 +257,34 @@ It should be based on sound mathematical, physical or computer science principle
 This is important for test code too. Tests shall be based on such principles and real-world properties so they don't just test the current implementation as it happens to be.
 
 Guidelines not tramlines: guidelines are not totally rigid - they can be broken when there is good reason.
+
+#Avoiding Accidental Mistakes
+
+* Always use braces when nesting if/else statements to avoid dangling else (the else branch belongs to the innermost if statement).
```
Comment (ledvinap) [inline docs/development/CodingStyle.md]: IMO BF decided to use blocks for `if`  in all cases, so this is a bit redundant.

(Personally I prefer to skip braces in some cases:
```
if (test)
  return;
if (test)
  continue;
```
)

File: docs/development/CodingStyle.md [hunk 3]
```diff
@@ -257,3 +257,34 @@ It should be based on sound mathematical, physical or computer science principle
...
+
+```cpp
+if ( var1 ) {
+	if ( var2 ) {
+		someMethod( var1 );
+	}
```
Comment (ledvinap) [inline docs/development/CodingStyle.md]:
```
} else {

File: docs/development/CodingStyle.md (hunk 2 above)
Comment (cpsw) [reply docs/development/CodingStyle.md]: I would prefer to use no braces in these two examples also. The rule is for nesting if/else statements, in which we might face dangling else problems. Thanks!

File: docs/development/CodingStyle.md (hunk 1 above)
Comment (cpsw) [reply docs/development/CodingStyle.md]: Sure, I can remove it.

File: docs/development/CodingStyle.md (hunk 2 above)
Comment (martinbudden) [reply docs/development/CodingStyle.md]: @cpsw , @ledvinap ,the reason for always using braces is to do with viewing differences. If braces are always used and extra code is added, then the difference between the old and the new code is just the changed code. Braces are not always used then the difference also contains extra lines to do with the braces.

Although having a few extra lines with braces may not seem much, it makes the diffs bigger and adds up, especially with large PRs. Code reviews are easier if the review is not distracted by irrelevant changes.

Comment (mikeller) [reply docs/development/CodingStyle.md]: I am with @cpsw on this. In terms of maintainability, and the fact that code changes over time, it is easy to get from this case to a hard to detect bug.

```
if (test)
#if defined(SPECIAL_CASE)
  done = true;
#endif
  continue;
```

**Comment by hydra** [comment]
can we get an updated PR based on the feedback please?

**Comment by hydra** [comment]
closing, stale, no activity.

**Pull request closed without merging**